### PR TITLE
Add code formatter

### PR DESCRIPTION
**Pull request by reacheight**
Implemented initial code formatter with some indent rules. For example, now 'Rename' refactoring keeps correct indents for multiline list value and match expressions.
ezgif-7-106f6b46c14e
match

### CODE REVIEW

File: ReSharper.FSharp/src/FSharp.Psi/src/CodeStyle/FSharpFormatSettingsKey.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
+using JetBrains.Application.Settings;
+using JetBrains.ReSharper.Psi.CodeStyle;
+using JetBrains.ReSharper.Psi.EditorConfig;
+using JetBrains.ReSharper.Psi.Format;
+
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.CodeStyle
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/CodeStyle/FSharpFormatSettingsKey.cs]: We should probably keep the old namespace here to retain existing Fantomas/code style settings.
I'm not sure if we have statistics of how many people have changed F# code style settings.

File: ReSharper.FSharp/src/FSharp.Psi/src/CodeStyle/FSharpFormatterInfoProvider.cs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.CodeStyle
+{
+  [Language(typeof(FSharpLanguage))]
+  public class FSharpFormatterInfoProvider :
+    FormatterInfoProviderWithFluentApi<CodeFormattingContext, FSharpFormatSettingsKey>
+  {
+    public FSharpFormatterInfoProvider(ISettingsSchema settingsSchema) : base(settingsSchema)
+    {
+      lock (this)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/CodeStyle/FSharpFormatterInfoProvider.cs]: Is there an actual need for the lock?
Please look at how `Language` components are created, find usages of `LanguageAttribute` in the SDK/product code.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/LanguageService/FSharpLanguageService.fs [hunk 3]
```diff
@@ -24,7 +24,7 @@ open JetBrains.Util
 [<Language(typeof<FSharpLanguage>)>]
 type FSharpLanguageService
         (languageType, constantValueService, cacheProvider: FSharpCacheProvider, checkerService: FSharpCheckerService,
-         formatter: FSharpDummyCodeFormatter) =
+         formatter: IFSharpCodeFormatter) =
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/LanguageService/FSharpLanguageService.fs]: What is the purpose of this interface? Is it going to be used in some way in future?

File: ReSharper.FSharp/src/FSharp.Psi/src/CodeStyle/FSharpFormatterInfoProvider.cs [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+      {
+        Describe<FormattingRule>()
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/CodeStyle/FSharpFormatterInfoProvider.cs]: It'd probably be neat to split the rules into several methods, like it's done for other languages. Maybe to have `Formatting` and `Indent` for now, and to regroup later if needed .

File: ReSharper.FSharp/src/FSharp.Psi/src/CodeStyle/FSharpCodeFormatter.cs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    public override void FormatReplacedNode(ITreeNode oldNode, ITreeNode newNode) =>
+      FormatInsertedNodes(newNode, newNode, false);
+
+    public override void FormatReplacedRange(ITreeNode first, ITreeNode last, ITreeRange oldNodes) =>
+      FormatInsertedNodes(first, last, false);
+
+    public override void FormatDeletedNodes(ITreeNode parent, ITreeNode prevNode, ITreeNode nextNode) =>
+      FormatterImplHelper.FormatDeletedNodesHelper(this, parent, prevNode, nextNode, false);
+
+    public override string OverridenSettingPrefix => "// @formatter:";
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/CodeStyle/FSharpCodeFormatter.cs]: Does it actually work? :)

**Comment by auduchinok** [comment]
If anyone is wondering why it's all written in C#, it's because of lack of a proper covariance/contravariance support in F#, and formatter rules types rely on it a lot. See https://github.com/dotnet/fsharp/issues/5277.

File: ReSharper.FSharp/src/FSharp.Psi/src/FSharp.psi [hunk 6]
```diff
@@ -528,13 +528,16 @@ forEachExpr options { stubBase="FSharpExpressionBase"; }:
   synExpr<DO_EXPR, DoExpression>;
 
 arrayOrListOfSeqExpr options { stubBase="FSharpExpressionBase"; }:
-  (LBRACK
-   synExpr<EXPR, Expression>
-   RBRACK) |
-
-   (LBRACK_BAR
+  (
+    LBRACK<LDELIMITER, LDelimiter>
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/FSharp.psi]: Maybe `LeftBracket` would be better?

File: ReSharper.FSharp/src/FSharp.Psi/src/CodeStyle/FSharpFormatterInfoProvider.cs (hunk 2 above)
Comment (reacheight) [reply ReSharper.FSharp/src/FSharp.Psi/src/CodeStyle/FSharpFormatterInfoProvider.cs]: Yes, there are asserts for locks in rules initialization.

Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi/src/CodeStyle/FSharpFormatterInfoProvider.cs]: OK, there are asserts for this kind of locks in the base types, I see.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Intentions/ToMultilineRecord.fs [hunk 7]
```diff
@@ -40,14 +41,15 @@ type ToMultilineRecord(dataProvider: FSharpContextActionDataProvider) =
...
+        // todo: rewrite using ModificationUtil when code formatter rules are ready
         for binding in bindings do
             if binding != firstBinding then
-                let newLine =
-                    match binding.PrevSibling with
-                    | Whitespace node -> ModificationUtil.ReplaceChild(node, NewLine(lineEnding))
-                    | node -> ModificationUtil.AddChildAfter(node, NewLine(lineEnding))
+                let newLine = NewLine(lineEnding)
+                match binding.PrevSibling with
+                | Whitespace node -> LowLevelModificationUtil.ReplaceChildRange(node, node, newLine)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/Intentions/ToMultilineRecord.fs]: Let's use `DisableCodeFormatter` or something similar instead until we're ready to use the formatter here?

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/GenerateMissingRecordFieldsFix.fs [hunk 8]
```diff
@@ -104,16 +105,18 @@ type GenerateMissingRecordFieldsFix(recordExpr: IRecordExpr) =
             | :? IRecordExprBinding -> anchor.Indent
             | _ -> anchor.Indent + 1
 
+        // todo: rewrite using only ModificationUtil when code formatter rules are ready
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/GenerateMissingRecordFieldsFix.fs]: Same here and in other places.

File: ReSharper.FSharp/src/FSharp.Psi/src/FSharp.psi [hunk 9]
```diff
@@ -543,10 +543,9 @@ private forEachInClause{ROLE, Getter}:
 
 arrayOrListOfSeqExpr options { stubBase="FSharpExpressionBase"; }:
   (LBRACK
-   synExpr<EXPR, Expression>
-   RBRACK) |
-
-   (LBRACK_BAR
+    synExpr<EXPR, Expression>
+    RBRACK) |
+  (LBRACK_BAR
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/FSharp.psi]: What does this change do?

File: ReSharper.FSharp/test/data/features/service/codeFormatter/Match lambda expr alignment 01 - Correct alignment.fs.gold [hunk 10]
```diff
@@ -0,0 +1,2 @@
+﻿let f = function
+        | _ -> ()
```
Comment (auduchinok) [inline ReSharper.FSharp/test/data/features/service/codeFormatter/Match lambda expr alignment 01 - Correct alignment.fs.gold]: Should add indent relatively the outer indentation:
```
let f = function
  | _ -> ()
```

It's good to check what indentation is allowed by looking at LexFilter module in the compiler.

**Pull request closed without merging**